From e726d3413fa66e2066138a7770ad9262b3cfe7ac Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 15 Mar 2024 15:37:39 +0100
Subject: [PATCH 0159/1010] net: dovetail: add infrastructure for oob device
 I/O

Enabling oob I/O for netdevices requires:

- the extension of the page pool implementation in order to enable oob
  access to dedicated pool of pages for buffer allocation and release
  (PP_FLAG_PAGE_OOB). This simplifies the addition of oob support to
  the netdevice drivers already using page pools for allocating their
  buffers (e.g. XDP-enabled drivers normally do).

- a significant rework of the Dovetail-specific skbuff handling in
  order to support proper recycling by the companion core, depending
  on the stage buffers are released from, and which among the in-band
  or oob netstacks actually owns them. The gist of the changes is
  about decoupling the information about the stage which owns the
  buffer head (skb->oob_head set if oob) from the information about
  the stage which owns the storage the buffer meta-data refers to
  (skb_shinfo(skb)->flags & SKBFL_OOB_STORAGE is non-zero if the
  storage lives in an oob-enabled page pool).

- the introduction of a global pool of buffer heads which may be
  accessed from the oob stage for allocation and release.

- making __build_skb() aware of the existence of the oob buffer pool,
  so that a new head from the latter when the caller runs oob, which
  eventually makes this routine oob-capable.

- teaching a critical portion of the NAPI how to deal with the
  out-of-band execution stage, so that fewer changes to NIC drivers
  are required for enabling oob traffic.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/dovetail/netdevice.h |  11 ++
 include/linux/net.h          |  10 ++
 include/linux/netdevice.h    | 123 +++++++++++----------
 include/linux/skbuff.h       | 191 ++++++++++++++++++++------------
 include/net/page_pool.h      |  28 ++++-
 net/Kconfig                  |   5 +
 net/core/dev.c               | 203 +++++++++++++++++++++++-----------
 net/core/gro.c               |   7 ++
 net/core/page_pool.c         |  93 +++++++++++++++-
 net/core/skbuff.c            | 207 +++++++++++++++++++----------------
 net/core/sysctl_net_core.c   |  11 ++
 net/ipv4/route.c             |   2 +-
 12 files changed, 597 insertions(+), 294 deletions(-)

diff --git a/include/dovetail/netdevice.h b/include/dovetail/netdevice.h
index 5bd975ccd5..2939e017bb 100644
--- a/include/dovetail/netdevice.h
+++ b/include/dovetail/netdevice.h
@@ -9,4 +9,15 @@
 struct oob_netdev_state {
 };
 
+struct oob_netqueue_state {
+};
+
+static inline void netqueue_init_oob(struct oob_netqueue_state *qs)
+{
+}
+
+static inline void netqueue_destroy_oob(struct oob_netqueue_state *qs)
+{
+}
+
 #endif /* !_DOVETAIL_NETDEVICE_H */
diff --git a/include/linux/net.h b/include/linux/net.h
index 17d055df8d..844742fe6b 100644
--- a/include/linux/net.h
+++ b/include/linux/net.h
@@ -326,6 +326,16 @@ static inline bool sendpage_ok(struct page *page)
 	return !PageSlab(page) && page_count(page) >= 1;
 }
 
+static inline bool net_running_oob(void)
+{
+	/*
+	 * Ensure depending code blocks are compiled out if
+	 * CONFIG_NET_OOB is disabled even if CONFIG_DOVETAIL is
+	 * enabled.
+	 */
+	return IS_ENABLED(CONFIG_NET_OOB) && running_oob();
+}
+
 int kernel_sendmsg(struct socket *sock, struct msghdr *msg, struct kvec *vec,
 		   size_t num, size_t len);
 int kernel_sendmsg_locked(struct sock *sk, struct msghdr *msg,
diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 85be5a4aac..84f509880d 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -327,6 +327,7 @@ enum netdev_state_t {
 	__LINK_STATE_LINKWATCH_PENDING,
 	__LINK_STATE_DORMANT,
 	__LINK_STATE_TESTING,
+	__LINK_STATE_OOB_PORT,
 	__LINK_STATE_OOB,
 };
 
@@ -472,6 +473,7 @@ static inline bool napi_prefer_busy_poll(struct napi_struct *n)
 }
 
 bool napi_schedule_prep(struct napi_struct *n);
+bool napi_schedule_unprep(struct napi_struct *n);
 
 /**
  *	napi_schedule - schedule NAPI poll
@@ -643,6 +645,8 @@ struct netdev_queue {
 
 	unsigned long		state;
 
+	struct oob_netqueue_state oob;
+
 #ifdef CONFIG_BQL
 	struct dql		dql;
 #endif
@@ -1399,18 +1403,16 @@ struct netdev_net_notifier {
  *	Get hardware timestamp based on normal/adjustable time or free running
  *	cycle counter. This function is required if physical clock supports a
  *	free running cycle counter.
- * struct sk_buff *(*ndo_alloc_oob_skb)(struct net_device *dev,
- * 					dma_addr_t *dma_addr);
- *	Allocate a buffer for out-of-band I/O. The memory must be immediately
- *	DMA-suitable, living in the linear kernel address space. i.e. no highmem.
- *	This handler must be provided if IFF_OOB_CAPABLE is set in the out-of-band
- *	context flags.
- * void	(*ndo_free_oob_skb)(struct net_device *dev,
- *			    struct sk_buff *skb,
- *			    dma_addr_t dma_addr);
- *	Free a buffer previously allocated from ndo_alloc_oob_skb() exclusively.
- *	This handler must be provided if IFF_OOB_CAPABLE is set in the out-of-band
- *	context flags.
+ * void	(*ndo_enable_oob)(struct net_device *dev);
+ *	Turn on out-of-band I/O handling. On return from this handler, the device
+ *	must be prepared to handle RX/TX packets from the out-of-band stage. This
+ *	handler is optional.
+ * void	(*ndo_disable_oob)(struct net_device *dev);
+ *	Turn off out-of-band I/O handling, reverting the effect of ndo_enable_oob().
+ *	This handler is optional.
+ * void	(*ndo_disable_oob)(struct net_device *dev);
+ *	Turn off out-of-band I/O handling, reverting the effect of ndo_enable_oob().
+ *	This handler is optional.
  */
 struct net_device_ops {
 	int			(*ndo_init)(struct net_device *dev);
@@ -1637,11 +1639,8 @@ struct net_device_ops {
 						  const struct skb_shared_hwtstamps *hwtstamps,
 						  bool cycles);
 #ifdef CONFIG_NET_OOB
-	struct sk_buff *	(*ndo_alloc_oob_skb)(struct net_device *dev,
-						     dma_addr_t *dma_addr);
-	void			(*ndo_free_oob_skb)(struct net_device *dev,
-						    struct sk_buff *skb,
-						    dma_addr_t dma_addr);
+	int			(*ndo_enable_oob)(struct net_device *dev);
+	void			(*ndo_disable_oob)(struct net_device *dev);
 #endif
 };
 
@@ -1694,7 +1693,6 @@ struct net_device_ops {
  *	skb_headlen(skb) == 0 (data starts from frag0)
  * @IFF_CHANGE_PROTO_DOWN: device supports setting carrier via IFLA_PROTO_DOWN
  * @IFF_OOB_CAPABLE: device supports direct out-of-band I/O operations.
- * @IFF_OOB_PORT: device is an active out-of-band port.
  */
 enum netdev_priv_flags {
 	IFF_802_1Q_VLAN			= 1<<0,
@@ -1731,7 +1729,6 @@ enum netdev_priv_flags {
 	IFF_TX_SKB_NO_LINEAR		= BIT_ULL(31),
 	IFF_CHANGE_PROTO_DOWN		= BIT_ULL(32),
 	IFF_OOB_CAPABLE			= BIT_ULL(33),
-	IFF_OOB_PORT			= BIT_ULL(34),
 };
 
 #define IFF_802_1Q_VLAN			IFF_802_1Q_VLAN
@@ -1767,7 +1764,6 @@ enum netdev_priv_flags {
 #define IFF_LIVE_RENAME_OK		IFF_LIVE_RENAME_OK
 #define IFF_TX_SKB_NO_LINEAR		IFF_TX_SKB_NO_LINEAR
 #define IFF_OOB_CAPABLE			IFF_OOB_CAPABLE
-#define IFF_OOB_PORT			IFF_OOB_PORT
 
 /* Specifies the type of the struct net_device::ml_priv pointer */
 enum netdev_ml_priv_type {
@@ -3196,6 +3192,11 @@ struct softnet_data {
 #endif
 #ifdef CONFIG_NET_FLOW_LIMIT
 	struct sd_flow_limit __rcu *flow_limit;
+#endif
+#ifdef CONFIG_NET_OOB
+	struct list_head	inband_rx_list; /* Inband skbs received oob */
+	hard_spinlock_t		inband_rx_lock;
+	struct irq_work		inband_rx_work;
 #endif
 	struct Qdisc		*output_queue;
 	struct Qdisc		**output_queue_tailp;
@@ -4302,14 +4303,19 @@ void netif_device_attach(struct net_device *dev);
 
 #ifdef CONFIG_NET_OOB
 
+static inline bool netdev_is_oob_capable(struct net_device *dev)
+{
+	return !!(dev->priv_flags & IFF_OOB_CAPABLE);
+}
+
 bool netif_receive_oob(struct sk_buff *skb);
 
 /**
- *	netif_oob_diversion - is I/O traffic diverted for out-of-band handling.
+ *	netif_oob_diversion - is the ingress traffic diverted for out-of-band handling?
  *	@dev: network device
  *
- * Check if the RX/TX packets are currently diverted for out-of-band
- * handling by a companion kernel.
+ * Check if the RX packets are currently diverted to a companion
+ * kernel for out-of-band handling.
  */
 static inline bool netif_oob_diversion(const struct net_device *dev)
 {
@@ -4318,51 +4324,58 @@ static inline bool netif_oob_diversion(const struct net_device *dev)
 
 static inline void netif_enable_oob_diversion(struct net_device *dev)
 {
-	return set_bit(__LINK_STATE_OOB, &dev->state);
+	const struct net_device_ops *ops = dev->netdev_ops;
+
+	if (ops->ndo_enable_oob)
+		ops->ndo_enable_oob(dev);
+
+	set_bit(__LINK_STATE_OOB, &dev->state);
 }
 
 static inline void netif_disable_oob_diversion(struct net_device *dev)
 {
+	const struct net_device_ops *ops = dev->netdev_ops;
+
 	clear_bit(__LINK_STATE_OOB, &dev->state);
 	smp_mb__after_atomic();
-}
-
-int netif_xmit_oob(struct sk_buff *skb);
-
-static inline bool netdev_is_oob_capable(struct net_device *dev)
-{
-	return !!(dev->priv_flags & IFF_OOB_CAPABLE);
-}
 
-static inline void netdev_enable_oob_port(struct net_device *dev)
-{
-	dev->priv_flags |= IFF_OOB_PORT;
+	if (ops->ndo_disable_oob)
+		ops->ndo_disable_oob(dev);
 }
 
-static inline void netdev_disable_oob_port(struct net_device *dev)
+/**
+ *	netif_oob_port - is the interface an active oob port?
+ *	@dev: network device
+ *
+ * Check whether the interface acts as a virtual port for handling oob
+ * traffic.
+ */
+static inline bool netif_oob_port(struct net_device *dev)
 {
-	dev->priv_flags &= ~IFF_OOB_PORT;
+	return test_bit(__LINK_STATE_OOB_PORT, &dev->state);
 }
 
-static inline bool netdev_is_oob_port(struct net_device *dev)
+static inline void netif_enable_oob_port(struct net_device *dev)
 {
-	return !!(dev->priv_flags & IFF_OOB_PORT);
+	set_bit(__LINK_STATE_OOB_PORT, &dev->state);
 }
 
-static inline struct sk_buff *netdev_alloc_oob_skb(struct net_device *dev,
-						   dma_addr_t *dma_addr)
+static inline void netif_disable_oob_port(struct net_device *dev)
 {
-	return dev->netdev_ops->ndo_alloc_oob_skb(dev, dma_addr);
+	clear_bit(__LINK_STATE_OOB_PORT, &dev->state);
+	smp_mb__after_atomic();
 }
 
-static inline void netdev_free_oob_skb(struct net_device *dev,
-				       struct sk_buff *skb,
-				       dma_addr_t dma_addr)
-{
-	dev->netdev_ops->ndo_free_oob_skb(dev, skb, dma_addr);
-}
+/* Out-of-band hooks implemented by the companion core. */
+void napi_schedule_oob(struct napi_struct *n);
+void napi_complete_oob(struct napi_struct *n);
+bool netif_deliver_oob(struct sk_buff *skb);
+int netif_xmit_oob(struct sk_buff *skb);
+void netif_tx_lock_oob(struct netdev_queue *txq);
+void netif_tx_unlock_oob(struct netdev_queue *txq);
+void process_inband_tx_backlog(struct softnet_data *sd);
 
-#else
+#else  /* !CONFIG_NET_OOB */
 
 static inline bool netif_receive_oob(struct sk_buff *skb)
 {
@@ -4374,25 +4387,25 @@ static inline bool netif_oob_diversion(const struct net_device *dev)
 	return false;
 }
 
-static inline bool netdev_is_oob_capable(struct net_device *dev)
+static inline bool netif_oob_port(struct net_device *dev)
 {
 	return false;
 }
 
-static inline void netdev_enable_oob_port(struct net_device *dev)
+static inline bool netdev_is_oob_capable(struct net_device *dev)
 {
+	return false;
 }
 
-static inline void netdev_disable_oob_port(struct net_device *dev)
+static inline void netif_tx_lock_oob(struct netdev_queue *txq)
 {
 }
 
-static inline bool netdev_is_oob_port(struct net_device *dev)
+static inline void netif_tx_unlock_oob(struct netdev_queue *txq)
 {
-	return false;
 }
 
-#endif
+#endif  /* !CONFIG_NET_OOB */
 
 /*
  * Network interface message level settings
diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 3ad11ccfd1..96eb54ff1c 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -525,6 +525,9 @@ enum {
 	 * use frags only up until ubuf_info is released
 	 */
 	SKBFL_MANAGED_FRAG_REFS = BIT(4),
+
+	/* data lives into an oob-enabled storage. */
+	SKBFL_OOB_STORAGE = BIT(5),
 };
 
 #define SKBFL_ZEROCOPY_FRAG	(SKBFL_ZEROCOPY_ENABLE | SKBFL_SHARED_FRAG)
@@ -924,14 +927,12 @@ struct sk_buff {
 				head_frag:1,
 				pfmemalloc:1,
 				pp_recycle:1; /* page_pool recycle indicator */
+#ifdef CONFIG_NET_OOB
+	__u8			oob_head:1;
+#endif
 #ifdef CONFIG_SKB_EXTENSIONS
 	__u8			active_extensions;
 #endif
-#ifdef CONFIG_NET_OOB
-	__u8			oob:1;
-	__u8			oob_clone:1;
-	__u8			oob_cloned:1;
-#endif
 
 	/* Fields enclosed in headers group are copied
 	 * using a single memcpy() in __copy_skb_header()
@@ -1262,70 +1263,6 @@ struct sk_buff *build_skb_around(struct sk_buff *skb,
 
 void skb_attempt_defer_free(struct sk_buff *skb);
 
-#ifdef CONFIG_NET_OOB
-
-static inline void __skb_oob_copy(struct sk_buff *new,
-				const struct sk_buff *old)
-{
-	new->oob = old->oob;
-	new->oob_clone = old->oob_clone;
-	new->oob_cloned = old->oob_cloned;
-}
-
-static inline bool skb_is_oob(const struct sk_buff *skb)
-{
-	return skb->oob;
-}
-
-static inline bool skb_is_oob_clone(const struct sk_buff *skb)
-{
-	return skb->oob_clone;
-}
-
-static inline bool skb_has_oob_clone(const struct sk_buff *skb)
-{
-	return skb->oob_cloned;
-}
-
-struct sk_buff *__netdev_alloc_oob_skb(struct net_device *dev,
-				size_t len, size_t headroom,
-				gfp_t gfp_mask);
-void __netdev_free_oob_skb(struct net_device *dev, struct sk_buff *skb);
-void netdev_reset_oob_skb(struct net_device *dev, struct sk_buff *skb,
-			size_t headroom);
-struct sk_buff *skb_alloc_oob_head(gfp_t gfp_mask);
-void skb_morph_oob_skb(struct sk_buff *n, struct sk_buff *skb);
-bool skb_release_oob_skb(struct sk_buff *skb, int *dref);
-
-static inline bool recycle_oob_skb(struct sk_buff *skb)
-{
-	bool skb_oob_recycle(struct sk_buff *skb);
-
-	if (!skb->oob)
-		return false;
-
-	return skb_oob_recycle(skb);
-}
-
-#else  /* !CONFIG_NET_OOB */
-
-static inline void __skb_oob_copy(struct sk_buff *new,
-				const struct sk_buff *old)
-{
-}
-
-static inline bool skb_is_oob(const struct sk_buff *skb)
-{
-	return false;
-}
-
-static inline bool recycle_oob_skb(struct sk_buff *skb)
-{
-	return false;
-}
-
-#endif	/* !CONFIG_NET_OOB */
-
 struct sk_buff *napi_build_skb(void *data, unsigned int frag_size);
 
 /**
@@ -5207,5 +5144,121 @@ static inline bool skb_pp_recycle(struct sk_buff *skb, void *data)
 	return page_pool_return_skb_page(virt_to_page(data));
 }
 
+#ifdef CONFIG_NET_OOB
+
+extern unsigned int sysctl_max_oob_skb;
+
+struct skbuff_oob_pool {
+	struct list_head pool;
+	hard_spinlock_t lock;
+};
+
+struct sk_buff *get_oob_skb(void);
+
+void put_oob_skb(struct sk_buff *skb);
+
+/**
+ * skb_is_oob - Whether the buffer head was allocated from the oob
+ * pool. Caution: this is distinct from the skb conveying oob storage,
+ * see skb_has_oob_storage().
+ */
+static inline bool skb_is_oob(const struct sk_buff *skb)
+{
+	return skb->oob_head;
+}
+
+static inline void skb_mark_oob(struct sk_buff *skb)
+{
+	skb->oob_head = 1;
+}
+
+/**
+ * skb_has_oob_storage - Whether the skb data lives in an oob-enabled
+ * page pool.
+ */
+static inline bool skb_has_oob_storage(const struct sk_buff *skb)
+{
+	return skb_shinfo(skb)->flags & SKBFL_OOB_STORAGE;
+}
+
+static inline void skb_mark_oob_storage(struct sk_buff *skb)
+{
+	skb_shinfo(skb)->flags |= SKBFL_OOB_STORAGE;
+}
+
+static inline dma_addr_t skb_oob_storage_addr(const struct sk_buff *skb)
+{
+	if (!skb->head || !skb_has_oob_storage(skb) || !skb->pp_recycle)
+		return DMA_MAPPING_ERROR;
+
+	return page_pool_get_dma_addr(virt_to_page(skb->head));
+}
+
+bool recycle_skb_oob(struct sk_buff *skb);
+void free_skb_oob(struct sk_buff *skb);
+void finalize_skb_inband(struct sk_buff *skb);
+
+/**
+ *	__skb_inband_clone - In-band specific setup for skbs for a
+ *	newly allocated clone head. Other allocation places don't need
+ *	that because they are memset-zeroed from skb to &skb->end
+ *	before building the buffer around a data storage area.
+ */
+static inline void __skb_inband_clone(struct sk_buff *skb)
+{
+	skb->oob_head = 0;
+}
+
+/**
+ *	__skb_oob_free_head - Called from the in-band net core after
+ *      the last reference to the buffer head (->users) was dropped,
+ *      and any (shared) data was unref'ed (and possibly freed). By
+ *      construction, this is only called for buffer heads which
+ *      refer(red) to non-oob storage. Buffers with oob heads _and_
+ *      storage must flow through free_skb_oob() instead.
+ */
+static inline bool __skb_oob_free_head(struct sk_buff *skb)
+{
+	if (skb_is_oob(skb)) {
+		put_oob_skb(skb);
+		return true;
+	}
+
+	return false;
+}
+
+#else  /* !CONFIG_NET_OOB */
+
+static inline bool skb_is_oob(const struct sk_buff *skb)
+{
+	return false;
+}
+
+static inline void __skb_inband_clone(struct sk_buff *skb)
+{
+}
+
+static inline bool __skb_oob_free_head(struct sk_buff *skb)
+{
+	return false;
+}
+
+static inline bool skb_has_oob_storage(const struct sk_buff *skb)
+{
+	return false;
+}
+
+static inline dma_addr_t skb_oob_storage_addr(const struct sk_buff *skb)
+{
+	return DMA_MAPPING_ERROR;
+}
+
+static inline bool recycle_skb_oob(struct sk_buff *skb)
+{
+	return false;
+}
+
+#endif	/* !CONFIG_NET_OOB */
+
 #endif	/* __KERNEL__ */
 #endif	/* _LINUX_SKBUFF_H */
diff --git a/include/net/page_pool.h b/include/net/page_pool.h
index ad0bafc877..db929c3640 100644
--- a/include/net/page_pool.h
+++ b/include/net/page_pool.h
@@ -46,9 +46,11 @@
 					* device driver responsibility
 					*/
 #define PP_FLAG_PAGE_FRAG	BIT(2) /* for page frag feature */
+#define PP_FLAG_PAGE_OOB	BIT(3) /* Enable access from the oob stage (Dovetail) */
 #define PP_FLAG_ALL		(PP_FLAG_DMA_MAP |\
 				 PP_FLAG_DMA_SYNC_DEV |\
-				 PP_FLAG_PAGE_FRAG)
+				 PP_FLAG_PAGE_FRAG |\
+				 PP_FLAG_PAGE_OOB)
 
 /*
  * Fast allocation side cache array/stack
@@ -64,12 +66,24 @@
  * cache is already full (or partly full) then the XDP_DROP recycles
  * would have to take a slower code path.
  */
-#define PP_ALLOC_CACHE_SIZE	128
-#define PP_ALLOC_CACHE_REFILL	64
+#ifdef CONFIG_PAGE_POOL_OOB
+#define PP_ALLOC_CACHE_SIZE(__pool)	((__pool)->p.pool_size)
+#define PP_ALLOC_CACHE_REFILL(__pool)	PP_ALLOC_CACHE_SIZE(__pool)
 struct pp_alloc_cache {
 	u32 count;
-	struct page *cache[PP_ALLOC_CACHE_SIZE];
+	struct page **cache;
+	hard_spinlock_t oob_lock;
 };
+#else	/* !CONFIG_PAGE_POOL_OOB */
+#define __PP_ALLOC_CACHE_SIZE	128
+#define __PP_ALLOC_CACHE_REFILL	64
+#define PP_ALLOC_CACHE_SIZE(__pool)	({ (void)__pool; __PP_ALLOC_CACHE_SIZE; })
+#define PP_ALLOC_CACHE_REFILL(__pool)	({ (void)__pool; __PP_ALLOC_CACHE_REFILL; })
+struct pp_alloc_cache {
+	u32 count;
+	struct page *cache[__PP_ALLOC_CACHE_SIZE];
+};
+#endif	/* !CONFIG_PAGE_POOL_OOB */
 
 struct page_pool_params {
 	unsigned int	flags;
@@ -239,6 +253,12 @@ inline enum dma_data_direction page_pool_get_dma_dir(struct page_pool *pool)
 	return pool->p.dma_dir;
 }
 
+static inline bool page_pool_is_oob(struct page_pool *pool)
+{
+	return IS_ENABLED(CONFIG_PAGE_POOL_OOB) &&
+		pool->p.flags & PP_FLAG_PAGE_OOB;
+}
+
 bool page_pool_return_skb_page(struct page *page);
 
 struct page_pool *page_pool_create(const struct page_pool_params *params);
diff --git a/net/Kconfig b/net/Kconfig
index 3b2fad8409..a0b464dfa4 100644
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -60,6 +60,8 @@ config SKB_EXTENSIONS
 
 config NET_OOB
 	bool
+	depends on DOVETAIL
+	select PAGE_POOL_OOB
 
 menu "Networking options"
 
@@ -435,6 +437,9 @@ config NET_DEVLINK
 config PAGE_POOL
 	bool
 
+config PAGE_POOL_OOB
+	def_bool n
+
 config PAGE_POOL_STATS
 	default n
 	bool "Page pool stats"
diff --git a/net/core/dev.c b/net/core/dev.c
index 9453518f9e..204db3f128 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3192,10 +3192,11 @@ void __dev_kfree_skb_irq(struct sk_buff *skb, enum skb_free_reason reason)
 		return;
 	}
 
-	if (recycle_oob_skb(skb))
+	get_kfree_skb_cb(skb)->reason = reason;
+
+	if (recycle_skb_oob(skb))
 		return;
 
-	get_kfree_skb_cb(skb)->reason = reason;
 	local_irq_save(flags);
 	skb->next = __this_cpu_read(softnet_data.completion_queue);
 	__this_cpu_write(softnet_data.completion_queue, skb);
@@ -3206,7 +3207,7 @@ EXPORT_SYMBOL(__dev_kfree_skb_irq);
 
 void __dev_kfree_skb_any(struct sk_buff *skb, enum skb_free_reason reason)
 {
-	if (in_hardirq() || irqs_disabled())
+	if (running_oob() || in_hardirq() || irqs_disabled())
 		__dev_kfree_skb_irq(skb, reason);
 	else if (unlikely(reason == SKB_REASON_DROPPED))
 		kfree_skb(skb);
@@ -3666,12 +3667,7 @@ static int xmit_one(struct sk_buff *skb, struct net_device *dev,
 	unsigned int len;
 	int rc;
 
-	/*
-	 * Clone-relay outgoing packet to listening taps. Network taps
-	 * interested in out-of-band traffic should be handled by the
-	 * companion core.
-	 */
-	if (dev_nit_active(dev) && !skb_is_oob(skb))
+	if (dev_nit_active(dev))
 		dev_queue_xmit_nit(skb, dev);
 
 	len = skb->len;
@@ -5027,7 +5023,9 @@ EXPORT_SYMBOL_GPL(do_xdp_generic);
 
 #ifdef CONFIG_NET_OOB
 
-__weak bool netif_oob_deliver(struct sk_buff *skb)
+static void _netif_receive_skb_list(struct list_head *head);
+
+__weak bool netif_deliver_oob(struct sk_buff *skb)
 {
 	return false;
 }
@@ -5039,44 +5037,71 @@ __weak int netif_xmit_oob(struct sk_buff *skb)
 
 bool netif_receive_oob(struct sk_buff *skb)
 {
+	struct softnet_data *sd = this_cpu_ptr(&softnet_data);
 	struct net_device *dev = skb->dev;
+	unsigned long flags;
+	bool kick_softirq;
 
-	if (dev && netif_oob_diversion(dev))
-		return netif_oob_deliver(skb);
+	/*
+	 * If the interface has oob diversion enabled, attempt to
+	 * deliver the incoming packet to the oob netstack. If the
+	 * packet is not picked, then decide whether we should leave
+	 * it to the in-band caller, or defer it to the RX softirq if
+	 * currently running oob.
+	 */
+	if (dev && netif_oob_diversion(dev)) {
+		if (netif_deliver_oob(skb))
+			return true;
+		if (net_running_oob()) {
+			raw_spin_lock_irqsave(&sd->inband_rx_lock, flags);
+			kick_softirq = list_empty(&sd->inband_rx_list);
+			list_add(&skb->list, &sd->inband_rx_list);
+			raw_spin_unlock_irqrestore(&sd->inband_rx_lock, flags);
+			if (kick_softirq)
+				irq_work_queue(&sd->inband_rx_work);
+			return true;
+		}
+	}
 
-	return false;
+	return false;	     /* Not handled, caller should pick it. */
 }
 
 static bool netif_receive_oob_list(struct list_head *head)
 {
 	struct sk_buff *skb, *next;
-	struct net_device *dev;
-
-	dev = list_first_entry(head, struct sk_buff, list)->dev;
-	if (!dev || !netif_oob_diversion(dev))
-		return false;
 
-	/* Callee dequeues every skb it consumes. */
 	list_for_each_entry_safe(skb, next, head, list)
-		netif_oob_deliver(skb);
+		netif_receive_oob(skb);
 
 	return list_empty(head);
 }
 
-__weak void netif_oob_run(struct net_device *dev)
-{ }
-
-static void napi_complete_oob(struct napi_struct *n)
+static void process_inband_rx_backlog(struct softnet_data *sd)
 {
-	struct net_device *dev = n->dev;
+	unsigned long flags;
+	LIST_HEAD(list);
 
-	if (netif_oob_diversion(dev))
-		netif_oob_run(dev);
+	raw_spin_lock_irqsave(&sd->inband_rx_lock, flags);
+	list_splice_init(&sd->inband_rx_list, &list);
+	raw_spin_unlock_irqrestore(&sd->inband_rx_lock, flags);
+	if (!list_empty(&list))
+		_netif_receive_skb_list(&list);
 }
 
-__weak void skb_inband_xmit_backlog(void)
+__weak void napi_schedule_oob(struct napi_struct *n)
 { }
 
+__weak void napi_complete_oob(struct napi_struct *n)
+{ }
+
+__weak void process_inband_tx_backlog(struct softnet_data *sd)
+{ }
+
+static void kick_rx_inband(struct irq_work *irq_work)
+{
+	__raise_softirq_irqoff(NET_RX_SOFTIRQ);
+}
+
 #else
 
 static inline bool netif_receive_oob_list(struct list_head *head)
@@ -5084,10 +5109,16 @@ static inline bool netif_receive_oob_list(struct list_head *head)
 	return false;
 }
 
+static inline void napi_schedule_oob(struct napi_struct *n)
+{ }
+
 static inline void napi_complete_oob(struct napi_struct *n)
 { }
 
-static inline void skb_inband_xmit_backlog(void)
+static inline void process_inband_tx_backlog(struct softnet_data *sd)
+{ }
+
+static void process_inband_rx_backlog(struct softnet_data *sd)
 { }
 
 #endif
@@ -5184,7 +5215,7 @@ static __latent_entropy void net_tx_action(struct softirq_action *h)
 {
 	struct softnet_data *sd = this_cpu_ptr(&softnet_data);
 
-	skb_inband_xmit_backlog();
+	process_inband_tx_backlog(sd);
 
 	if (sd->completion_queue) {
 		struct sk_buff *clist;
@@ -5925,6 +5956,18 @@ int netif_receive_skb(struct sk_buff *skb)
 }
 EXPORT_SYMBOL(netif_receive_skb);
 
+static void _netif_receive_skb_list(struct list_head *head)
+{
+	struct sk_buff *skb;
+
+	if (trace_netif_receive_skb_list_entry_enabled()) {
+		list_for_each_entry(skb, head, list)
+			trace_netif_receive_skb_list_entry(skb);
+	}
+	netif_receive_skb_list_internal(head);
+	trace_netif_receive_skb_list_exit(0);
+}
+
 /**
  *	netif_receive_skb_list - process many receive buffers from network
  *	@head: list of skbs to process.
@@ -5937,18 +5980,11 @@ EXPORT_SYMBOL(netif_receive_skb);
  */
 void netif_receive_skb_list(struct list_head *head)
 {
-	struct sk_buff *skb;
-
 	if (list_empty(head))
 		return;
 	if (netif_receive_oob_list(head))
 		return;
-	if (trace_netif_receive_skb_list_entry_enabled()) {
-		list_for_each_entry(skb, head, list)
-			trace_netif_receive_skb_list_entry(skb);
-	}
-	netif_receive_skb_list_internal(head);
-	trace_netif_receive_skb_list_exit(0);
+	_netif_receive_skb_list(head);
 }
 EXPORT_SYMBOL(netif_receive_skb_list);
 
@@ -6139,14 +6175,20 @@ static int process_backlog(struct napi_struct *napi, int quota)
  *
  * The entry's receive function will be scheduled to run.
  * Consider using __napi_schedule_irqoff() if hard irqs are masked.
+ *
+ * Dovetail: may be called from the out-of-band stage.
  */
 void __napi_schedule(struct napi_struct *n)
 {
 	unsigned long flags;
 
-	local_irq_save(flags);
-	____napi_schedule(this_cpu_ptr(&softnet_data), n);
-	local_irq_restore(flags);
+	if (net_running_oob()) {
+		napi_schedule_oob(n);
+	} else {
+		local_irq_save(flags);
+		____napi_schedule(this_cpu_ptr(&softnet_data), n);
+		local_irq_restore(flags);
+	}
 }
 EXPORT_SYMBOL(__napi_schedule);
 
@@ -6158,6 +6200,8 @@ EXPORT_SYMBOL(__napi_schedule);
  * it as running.  This is used as a condition variable to
  * insure only one NAPI poll instance runs.  We also make
  * sure there is no pending NAPI disable.
+ *
+ * Dovetail: may be called from the out-of-band stage.
  */
 bool napi_schedule_prep(struct napi_struct *n)
 {
@@ -6195,19 +6239,54 @@ EXPORT_SYMBOL(napi_schedule_prep);
  */
 void __napi_schedule_irqoff(struct napi_struct *n)
 {
-	if (!IS_ENABLED(CONFIG_PREEMPT_RT))
+	if (net_running_oob())
+		napi_schedule_oob(n);
+	else if (!IS_ENABLED(CONFIG_PREEMPT_RT))
 		____napi_schedule(this_cpu_ptr(&softnet_data), n);
 	else
 		__napi_schedule(n);
 }
 EXPORT_SYMBOL(__napi_schedule_irqoff);
 
+bool napi_schedule_unprep(struct napi_struct *n)
+{
+	unsigned long val, new;
+
+	do {
+		val = READ_ONCE(n->state);
+
+		WARN_ON_ONCE(!(val & NAPIF_STATE_SCHED));
+
+		new = val & ~(NAPIF_STATE_MISSED | NAPIF_STATE_SCHED |
+			      NAPIF_STATE_SCHED_THREADED |
+			      NAPIF_STATE_PREFER_BUSY_POLL);
+
+		/* If STATE_MISSED was set, leave STATE_SCHED set,
+		 * because we will call napi->poll() one more time.
+		 * This C code was suggested by Alexander Duyck to help gcc.
+		 */
+		new |= (val & NAPIF_STATE_MISSED) / NAPIF_STATE_MISSED *
+						    NAPIF_STATE_SCHED;
+	} while (cmpxchg(&n->state, val, new) != val);
+
+	if (unlikely(val & NAPIF_STATE_MISSED)) {
+		__napi_schedule(n);
+		return false;
+	}
+
+	return true;
+}
+
 bool napi_complete_done(struct napi_struct *n, int work_done)
 {
-	unsigned long flags, val, new, timeout = 0;
+	unsigned long flags, timeout = 0;
 	bool ret = true;
 
-	napi_complete_oob(n);
+	if (netif_oob_diversion(n->dev)) {
+		napi_complete_oob(n);
+		if (net_running_oob())
+			return true;
+	}
 
 	/*
 	 * 1) Don't let napi dequeue from the cpu poll list
@@ -6247,27 +6326,8 @@ bool napi_complete_done(struct napi_struct *n, int work_done)
 		local_irq_restore(flags);
 	}
 
-	do {
-		val = READ_ONCE(n->state);
-
-		WARN_ON_ONCE(!(val & NAPIF_STATE_SCHED));
-
-		new = val & ~(NAPIF_STATE_MISSED | NAPIF_STATE_SCHED |
-			      NAPIF_STATE_SCHED_THREADED |
-			      NAPIF_STATE_PREFER_BUSY_POLL);
-
-		/* If STATE_MISSED was set, leave STATE_SCHED set,
-		 * because we will call napi->poll() one more time.
-		 * This C code was suggested by Alexander Duyck to help gcc.
-		 */
-		new |= (val & NAPIF_STATE_MISSED) / NAPIF_STATE_MISSED *
-						    NAPIF_STATE_SCHED;
-	} while (cmpxchg(&n->state, val, new) != val);
-
-	if (unlikely(val & NAPIF_STATE_MISSED)) {
-		__napi_schedule(n);
+	if (!napi_schedule_unprep(n))
 		return false;
-	}
 
 	if (timeout)
 		hrtimer_start(&n->timer, ns_to_ktime(timeout),
@@ -6837,6 +6897,8 @@ static __latent_entropy void net_rx_action(struct softirq_action *h)
 	LIST_HEAD(list);
 	LIST_HEAD(repoll);
 
+	process_inband_rx_backlog(sd);
+
 	local_irq_disable();
 	list_splice_init(&sd->poll_list, &list);
 	local_irq_enable();
@@ -10078,13 +10140,21 @@ static void netdev_init_one_queue(struct net_device *dev,
 	queue->xmit_lock_owner = -1;
 	netdev_queue_numa_node_write(queue, NUMA_NO_NODE);
 	queue->dev = dev;
+	netqueue_init_oob(&queue->oob);
 #ifdef CONFIG_BQL
 	dql_init(&queue->dql, HZ);
 #endif
 }
 
+static inline void destroy_oob_queue(struct net_device *dev,
+				struct netdev_queue *nq, void *arg)
+{
+	netqueue_destroy_oob(&nq->oob);
+}
+
 static void netif_free_tx_queues(struct net_device *dev)
 {
+	netdev_for_each_tx_queue(dev, destroy_oob_queue, NULL);
 	kvfree(dev->_tx);
 }
 
@@ -11659,6 +11729,11 @@ static int __init net_dev_init(void)
 #ifdef CONFIG_RPS
 		INIT_CSD(&sd->csd, rps_trigger_softirq, sd);
 		sd->cpu = i;
+#endif
+#ifdef CONFIG_NET_OOB
+		INIT_LIST_HEAD(&sd->inband_rx_list);
+		raw_spin_lock_init(&sd->inband_rx_lock);
+		init_irq_work(&sd->inband_rx_work, kick_rx_inband);
 #endif
 		INIT_CSD(&sd->defer_csd, trigger_rx_softirq, sd);
 		spin_lock_init(&sd->defer_lock);
diff --git a/net/core/gro.c b/net/core/gro.c
index c986d47442..eba02df117 100644
--- a/net/core/gro.c
+++ b/net/core/gro.c
@@ -650,6 +650,13 @@ gro_result_t napi_gro_receive(struct napi_struct *napi, struct sk_buff *skb)
 {
 	gro_result_t ret;
 
+	/*
+	 * Note: GRO is bypassed for devices enabling out-of-band RX,
+	 * including for in-band traffic not picked by the companion
+	 * core. In this case, the traffic flows through
+	 * process_inband_rx_backlog() -> _netif_receive_skb_list()
+	 * via NET_RX_SOFTIRQ.
+	 */
 	if (netif_receive_oob(skb))
 		return GRO_NORMAL;
 
diff --git a/net/core/page_pool.c b/net/core/page_pool.c
index acc1d0d055..15bce7a0b2 100644
--- a/net/core/page_pool.c
+++ b/net/core/page_pool.c
@@ -156,6 +156,10 @@ static void page_pool_producer_unlock(struct page_pool *pool,
 		spin_unlock_bh(&pool->ring.producer_lock);
 }
 
+noinline
+static struct page *__page_pool_alloc_pages_slow(struct page_pool *pool,
+						gfp_t gfp);
+
 static int page_pool_init(struct page_pool *pool,
 			  const struct page_pool_params *params)
 {
@@ -169,6 +173,9 @@ static int page_pool_init(struct page_pool *pool,
 
 	if (pool->p.pool_size)
 		ring_qsize = pool->p.pool_size;
+	/* Size of oob-accessed pools must be specified. */
+	else if (page_pool_is_oob(pool))
+		return -EINVAL;
 
 	/* Sanity limit mem that can be pinned down */
 	if (ring_qsize > 32768)
@@ -203,6 +210,11 @@ static int page_pool_init(struct page_pool *pool,
 	    pool->p.flags & PP_FLAG_PAGE_FRAG)
 		return -EINVAL;
 
+	/* Cannot manage page frags in oob-accessed pools. */
+	if (page_pool_is_oob(pool) &&
+		pool->p.flags & PP_FLAG_PAGE_FRAG)
+		return -EINVAL;
+
 #ifdef CONFIG_PAGE_POOL_STATS
 	pool->recycle_stats = alloc_percpu(struct page_pool_recycle_stats);
 	if (!pool->recycle_stats)
@@ -224,6 +236,23 @@ static int page_pool_init(struct page_pool *pool,
 	if (pool->p.flags & PP_FLAG_DMA_MAP)
 		get_device(pool->p.dev);
 
+#ifdef CONFIG_PAGE_POOL_OOB
+	pool->alloc.cache = kzalloc_node(
+		sizeof(struct page *) * params->pool_size,
+		GFP_KERNEL, params->nid);
+	if (pool->alloc.cache == NULL) {
+#ifdef CONFIG_PAGE_POOL_STATS
+		free_percpu(pool->recycle_stats);
+#endif
+		ptr_ring_cleanup(&pool->ring, NULL);
+		return -ENOMEM;
+	}
+	raw_spin_lock_init(&pool->alloc.oob_lock);
+	/* Populate the fast cache of oob-accessed pools at init. */
+	if (page_pool_is_oob(pool))
+		__page_pool_alloc_pages_slow(pool, GFP_KERNEL);
+#endif
+
 	return 0;
 }
 
@@ -291,7 +320,7 @@ static struct page *page_pool_refill_alloc_cache(struct page_pool *pool)
 			page = NULL;
 			break;
 		}
-	} while (pool->alloc.count < PP_ALLOC_CACHE_REFILL);
+	} while (pool->alloc.count < PP_ALLOC_CACHE_REFILL(pool));
 
 	/* Return last page */
 	if (likely(pool->alloc.count > 0)) {
@@ -313,6 +342,9 @@ static struct page *__page_pool_get_cached(struct page_pool *pool)
 		page = pool->alloc.cache[--pool->alloc.count];
 		alloc_stat_inc(pool, fast);
 	} else {
+		/* Cache refilling is disabled for oob-accessed pools. */
+		if (unlikely(page_pool_is_oob(pool)))
+			return NULL;
 		page = page_pool_refill_alloc_cache(pool);
 	}
 
@@ -399,7 +431,7 @@ noinline
 static struct page *__page_pool_alloc_pages_slow(struct page_pool *pool,
 						 gfp_t gfp)
 {
-	const int bulk = PP_ALLOC_CACHE_REFILL;
+	const int bulk = PP_ALLOC_CACHE_REFILL(pool);
 	unsigned int pp_flags = pool->p.flags;
 	unsigned int pp_order = pool->p.order;
 	struct page *page;
@@ -414,7 +446,7 @@ static struct page *__page_pool_alloc_pages_slow(struct page_pool *pool,
 		return pool->alloc.cache[--pool->alloc.count];
 
 	/* Mark empty alloc.cache slots "empty" for alloc_pages_bulk_array */
-	memset(&pool->alloc.cache, 0, sizeof(void *) * bulk);
+	memset(pool->alloc.cache, 0, sizeof(void *) * bulk);
 
 	nr_pages = alloc_pages_bulk_array_node(gfp, pool->p.nid, bulk,
 					       pool->alloc.cache);
@@ -460,6 +492,27 @@ struct page *page_pool_alloc_pages(struct page_pool *pool, gfp_t gfp)
 	struct page *page;
 
 	/* Fast-path: Get a page from cache */
+
+#ifdef CONFIG_PAGE_POOL_OOB
+	if (page_pool_is_oob(pool)) {
+		unsigned long flags;
+		/*
+		 * Fast, but still serialized. Out-of-band NAPI
+		 * context would not protect us from races since we
+		 * might have in-band vs oob races too.
+		 */
+		raw_spin_lock_irqsave(&pool->alloc.oob_lock, flags);
+		page = __page_pool_get_cached(pool);
+		raw_spin_unlock_irqrestore(&pool->alloc.oob_lock, flags);
+		/*
+		 * Slow path is disabled for oob-accessed pools,
+		 * either we did find a free page in the cache, or
+		 * things are about to get ugly.
+		 */
+		return page;
+	}
+#endif
+
 	page = __page_pool_get_cached(pool);
 	if (page)
 		return page;
@@ -560,7 +613,7 @@ static bool page_pool_recycle_in_ring(struct page_pool *pool, struct page *page)
 static bool page_pool_recycle_in_cache(struct page *page,
 				       struct page_pool *pool)
 {
-	if (unlikely(pool->alloc.count == PP_ALLOC_CACHE_SIZE)) {
+	if (unlikely(pool->alloc.count == PP_ALLOC_CACHE_SIZE(pool))) {
 		recycle_stat_inc(pool, cache_full);
 		return false;
 	}
@@ -581,6 +634,35 @@ static __always_inline struct page *
 __page_pool_put_page(struct page_pool *pool, struct page *page,
 		     unsigned int dma_sync_size, bool allow_direct)
 {
+#ifdef CONFIG_PAGE_POOL_OOB
+	/*
+	 * Pages from oob-accessed pools always live in the fast
+	 * cache, never in the ring. They are released to the global
+	 * page allocator only when the pool is destroyed.
+	 */
+	if (page_pool_is_oob(pool)) {
+		unsigned long flags;
+		bool ret;
+		/*
+		 * If a page from an oob-accessed pool is still
+		 * referenced when released or the fast cache would
+		 * overflow, we bark at the console then bluntly leak
+		 * it. This situation should never occur, or would
+		 * denote a serious API usage issue anyway.
+		 */
+		if (WARN_ON(dovetail_debug() && page_ref_count(page) != 1))
+			return NULL;
+		if (pool->p.flags & PP_FLAG_DMA_SYNC_DEV)
+			page_pool_dma_sync_for_device(pool, page,
+						      dma_sync_size);
+		raw_spin_lock_irqsave(&pool->alloc.oob_lock, flags);
+		ret = page_pool_recycle_in_cache(page, pool);
+		raw_spin_unlock_irqrestore(&pool->alloc.oob_lock, flags);
+		WARN_ON(!ret);
+		return NULL;
+	}
+#endif
+
 	/* This allocator is optimized for the XDP mode that uses
 	 * one-frame-per-page, but have fallbacks that act like the
 	 * regular page allocator APIs.
@@ -790,6 +872,9 @@ static void page_pool_free(struct page_pool *pool)
 
 #ifdef CONFIG_PAGE_POOL_STATS
 	free_percpu(pool->recycle_stats);
+#endif
+#ifdef CONFIG_PAGE_POOL_OOB
+	kfree(pool->alloc.cache);
 #endif
 	kfree(pool);
 }
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index e5ef8f3092..11efdeed4f 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -298,113 +298,112 @@ static void __build_skb_around(struct sk_buff *skb, void *data,
 
 #ifdef CONFIG_NET_OOB
 
-struct sk_buff *__netdev_alloc_oob_skb(struct net_device *dev, size_t len,
-				size_t headroom, gfp_t gfp_mask)
-{
-	struct sk_buff *skb;
-
-	headroom = ALIGN(NET_SKB_PAD + headroom, NET_SKB_PAD);
-	skb = __alloc_skb(len + headroom, gfp_mask,
-			SKB_ALLOC_RX, NUMA_NO_NODE);
-	if (!skb)
-		return NULL;
+unsigned int sysctl_max_oob_skb __read_mostly = 1024;
+EXPORT_SYMBOL(sysctl_max_oob_skb);
 
-	skb_reserve(skb, headroom);
-	skb->dev = dev;
-	skb->oob = true;
-	skb->oob_clone = false;
-	skb->oob_cloned = false;
+__weak void free_skb_oob(struct sk_buff *skb)
+{ }
 
-	return skb;
-}
-EXPORT_SYMBOL_GPL(__netdev_alloc_oob_skb);
+__weak void free_skb_head_oob(struct sk_buff *skb)
+{ }
 
-void __netdev_free_oob_skb(struct net_device *dev, struct sk_buff *skb)
+bool recycle_skb_oob(struct sk_buff *skb)
 {
-	skb->oob = false;
-	skb->oob_clone = false;
-	skb->oob_cloned = false;
-	dev_kfree_skb(skb);
+	/*
+	 * Hand the buffer release over the out-of-band core either if
+	 * the latter owns the skb data or we are currently running
+	 * oob.
+	 */
+	if (running_oob() || skb_has_oob_storage(skb)) {
+		free_skb_oob(skb);
+		return true;
+	}
+
+	return false;
 }
-EXPORT_SYMBOL_GPL(__netdev_free_oob_skb);
 
-void netdev_reset_oob_skb(struct net_device *dev, struct sk_buff *skb,
-			size_t headroom)
+void finalize_skb_inband(struct sk_buff *skb)
 {
-	unsigned char *data = skb->head; /* Always from kmalloc_reserve(). */
-
-	if (WARN_ON_ONCE(!skb->oob || skb->oob_clone))
-		return;
-
-	memset(skb, 0, offsetof(struct sk_buff, tail));
-	__build_skb_around(skb, data, 0);
-	headroom = ALIGN(NET_SKB_PAD + headroom, NET_SKB_PAD);
-	skb_reserve(skb, headroom);
-	skb->oob = true;
-	skb->oob_clone = false;
-	skb->oob_cloned = false;
-	skb->dev = dev;
+	if (skb->fclone != SKB_FCLONE_UNAVAILABLE)
+		__kfree_skb(skb);
+	else
+		__kfree_skb_defer(skb);
 }
-EXPORT_SYMBOL_GPL(netdev_reset_oob_skb);
 
-struct sk_buff *skb_alloc_oob_head(gfp_t gfp_mask)
-{
-	struct sk_buff *skb = kmem_cache_alloc(skbuff_head_cache, gfp_mask);
+/*
+ * Pool for out-of-band allocation of buffer heads. The implementation
+ * is trivial ATM, we may improve this with per-CPU caches in the
+ * future.
+ */
+static struct skbuff_oob_pool skbuff_oob_pool;
 
-	if (!skb)
-		return NULL;
+void init_oob_cache(void)
+{
+	struct skbuff_oob_pool *c = &skbuff_oob_pool;
+	unsigned int n, max_skbs;
+	struct sk_buff *skb;
 
 	/*
-	 * skb heads allocated for out-of-band traffic should be
-	 * reserved for clones, so memset is extraneous in the sense
-	 * that skb_morph_oob() should follow the allocation.
+	 * Populate the oob pool with buffers pulled from the regular
+	 * head cache. Note: there is no point in using a llist for
+	 * maintaining the entries, our MPMC model requires locking,
+	 * so adding more atomic ops in the picture in addition to
+	 * spinlocking would make no sense.
 	 */
-	memset(skb, 0, offsetof(struct sk_buff, tail));
-	refcount_set(&skb->users, 1);
-	skb->oob = true;
-	skb->oob_clone = true;
-	skb->oob_cloned = false;
-	skb_set_kcov_handle(skb, kcov_common_handle());
+	INIT_LIST_HEAD(&c->pool);
+	raw_spin_lock_init(&c->lock);
 
-	return skb;
+	max_skbs = READ_ONCE(sysctl_max_oob_skb);
+	for (n = 0; n < max_skbs; n++) {
+		skb = kmem_cache_alloc(skbuff_head_cache, GFP_KERNEL);
+		BUG_ON(!skb);
+		list_add(&skb->list, &c->pool);
+	}
 }
-EXPORT_SYMBOL_GPL(skb_alloc_oob_head);
 
-static struct sk_buff *__skb_clone(struct sk_buff *n, struct sk_buff *skb);
-
-void skb_morph_oob_skb(struct sk_buff *n, struct sk_buff *skb)
+struct sk_buff *get_oob_skb(void)
 {
-	__skb_clone(n, skb);
-	n->oob = true;
-	n->oob_clone = true;
-	n->oob_cloned = true;
-	skb->oob_cloned = true;
+	struct skbuff_oob_pool *c = &skbuff_oob_pool;
+	struct sk_buff *skb = NULL;
+	unsigned long flags;
+
+	raw_spin_lock_irqsave(&c->lock, flags);
+	if (!list_empty(&c->pool)) {
+		skb = list_first_entry(&c->pool, struct sk_buff, list);
+		list_del(&skb->list);
+		raw_spin_unlock_irqrestore(&c->lock, flags);
+		memset(skb, 0, offsetof(struct sk_buff, tail));
+		skb_mark_oob(skb);
+	} else {
+		raw_spin_unlock_irqrestore(&c->lock, flags);
+	}
+
+	return skb;
 }
-EXPORT_SYMBOL_GPL(skb_morph_oob_skb);
+EXPORT_SYMBOL(get_oob_skb);
 
-bool skb_release_oob_skb(struct sk_buff *skb, int *dref)
+void put_oob_skb(struct sk_buff *skb)
 {
-	struct skb_shared_info *shinfo = skb_shinfo(skb);
+	struct skbuff_oob_pool *c = &skbuff_oob_pool;
+	unsigned long flags;
 
-	if (!skb_unref(skb))
-		return false;
+	raw_spin_lock_irqsave(&c->lock, flags);
+	list_add(&skb->list, &c->pool);
+	raw_spin_unlock_irqrestore(&c->lock, flags);
+}
+EXPORT_SYMBOL(put_oob_skb);
 
-	/*
-	 * ->nohdr is never set for oob shells, so we always refcount
-         * the full data (header + payload) when cloned.
-	 */
-	*dref = skb->cloned ? atomic_sub_return(1, &shinfo->dataref) : 0;
+#else  /* !CONFIG_NET_OOB */
 
-	return true;
-}
-EXPORT_SYMBOL_GPL(skb_release_oob_skb);
+static inline void init_oob_cache(void)
+{ }
 
-__weak bool skb_oob_recycle(struct sk_buff *skb)
+static inline struct sk_buff *get_oob_skb(void)
 {
-	return false;
+	return NULL;
 }
 
-#endif	/* CONFIG_NET_OOB */
+#endif	/* !CONFIG_NET_OOB */
 
 /**
  * __build_skb - build a network buffer
@@ -424,16 +423,25 @@ __weak bool skb_oob_recycle(struct sk_buff *skb)
  *  After IO, driver calls build_skb(), to allocate sk_buff and populate it
  *  before giving packet to stack.
  *  RX rings only contains data buffers, not full skbs.
+ *
+ * Dovetail: allocation requests issued from the oob execution stage
+ * are served by the oob buffer cache.
  */
 struct sk_buff *__build_skb(void *data, unsigned int frag_size)
 {
 	struct sk_buff *skb;
 
-	skb = kmem_cache_alloc(skbuff_head_cache, GFP_ATOMIC);
-	if (unlikely(!skb))
-		return NULL;
+	if (running_oob()) {
+		skb = get_oob_skb();
+		if (unlikely(!skb))
+			return NULL;
+	} else {
+		skb = kmem_cache_alloc(skbuff_head_cache, GFP_ATOMIC);
+		if (unlikely(!skb))
+			return NULL;
+		memset(skb, 0, offsetof(struct sk_buff, tail));
+	}
 
-	memset(skb, 0, offsetof(struct sk_buff, tail));
 	__build_skb_around(skb, data, frag_size);
 
 	return skb;
@@ -977,11 +985,12 @@ static void skb_release_all(struct sk_buff *skb)
 
 void __kfree_skb(struct sk_buff *skb)
 {
-	if (recycle_oob_skb(skb))
+	if (recycle_skb_oob(skb))
 		return;
 
 	skb_release_all(skb);
-	kfree_skbmem(skb);
+	if (!__skb_oob_free_head(skb))
+		kfree_skbmem(skb);
 }
 EXPORT_SYMBOL(__kfree_skb);
 
@@ -1185,11 +1194,12 @@ static void napi_skb_cache_put(struct sk_buff *skb)
 
 void __kfree_skb_defer(struct sk_buff *skb)
 {
-	if (recycle_oob_skb(skb))
+	if (recycle_skb_oob(skb))
 		return;
 
 	skb_release_all(skb);
-	napi_skb_cache_put(skb);
+	if (!__skb_oob_free_head(skb))
+		napi_skb_cache_put(skb);
 }
 
 void napi_skb_free_stolen_head(struct sk_buff *skb)
@@ -1215,7 +1225,7 @@ void napi_consume_skb(struct sk_buff *skb, int budget)
 		return;
 	}
 
-	DEBUG_NET_WARN_ON_ONCE(!in_softirq());
+	DEBUG_NET_WARN_ON_ONCE(running_inband() && !in_softirq());
 
 	if (!skb_unref(skb))
 		return;
@@ -1229,11 +1239,7 @@ void napi_consume_skb(struct sk_buff *skb, int budget)
 		return;
 	}
 
-	if (recycle_oob_skb(skb))
-		return;
-
-	skb_release_all(skb);
-	napi_skb_cache_put(skb);
+	__kfree_skb_defer(skb);
 }
 EXPORT_SYMBOL(napi_consume_skb);
 
@@ -1251,7 +1257,6 @@ static void __copy_skb_header(struct sk_buff *new, const struct sk_buff *old)
 	skb_dst_copy(new, old);
 	__skb_ext_copy(new, old);
 	__nf_copy(new, old, false);
-	__skb_oob_copy(new, old);
 
 	/* Note : this field could be in the headers group.
 	 * It is not yet because we do not want to have a 16 bit hole
@@ -1787,6 +1792,8 @@ struct sk_buff *skb_clone(struct sk_buff *skb, gfp_t gfp_mask)
 		n->fclone = SKB_FCLONE_UNAVAILABLE;
 	}
 
+	__skb_inband_clone(n);
+
 	return __skb_clone(n, skb);
 }
 EXPORT_SYMBOL(skb_clone);
@@ -4686,6 +4693,8 @@ void __init skb_init(void)
 						0,
 						SLAB_HWCACHE_ALIGN|SLAB_PANIC,
 						NULL);
+	init_oob_cache();
+
 	skb_extensions_init();
 }
 
@@ -5546,8 +5555,11 @@ EXPORT_SYMBOL(__skb_warn_lro_forwarding);
 void kfree_skb_partial(struct sk_buff *skb, bool head_stolen)
 {
 	if (head_stolen) {
-		skb_release_head_state(skb);
-		kmem_cache_free(skbuff_head_cache, skb);
+		if (!recycle_skb_oob(skb)) {
+			skb_release_head_state(skb);
+			if (!__skb_oob_free_head(skb))
+				kmem_cache_free(skbuff_head_cache, skb);
+		}
 	} else {
 		__kfree_skb(skb);
 	}
@@ -6512,6 +6524,7 @@ static int pskb_carve_inside_nonlinear(struct sk_buff *skb, const u32 off,
 	skb->len -= off;
 	skb->data_len = skb->len;
 	atomic_set(&skb_shinfo(skb)->dataref, 1);
+
 	return 0;
 }
 
diff --git a/net/core/sysctl_net_core.c b/net/core/sysctl_net_core.c
index 75efc712bb..69f41cf0b0 100644
--- a/net/core/sysctl_net_core.c
+++ b/net/core/sysctl_net_core.c
@@ -607,6 +607,17 @@ static struct ctl_table net_core_table[] = {
 		.proc_handler	= proc_dointvec_minmax,
 		.extra1		= SYSCTL_ZERO,
 	},
+#ifdef CONFIG_NET_OOB
+	{
+		.procname	= "max_oob_skb",
+		.data		= &sysctl_max_oob_skb,
+		.maxlen		= sizeof(unsigned int),
+		.mode		= 0444,
+		.proc_handler	= proc_dointvec_minmax,
+		.extra1		= SYSCTL_ONE_HUNDRED,
+		.extra2		= SYSCTL_INT_MAX,
+	},
+#endif
 	{ }
 };
 
diff --git a/net/ipv4/route.c b/net/ipv4/route.c
index e4be7ab208..0e1c7ee0f8 100644
--- a/net/ipv4/route.c
+++ b/net/ipv4/route.c
@@ -2926,7 +2926,7 @@ struct rtable *ip_route_output_flow(struct net *net, struct flowi4 *flp4,
 	 * core might want to learn about the routing decision, pass
 	 * it on.
 	 */
-	if (!IS_ERR(rt) && netdev_is_oob_port(rt->dst.dev))
+	if (!IS_ERR(rt) && netif_oob_port(rt->dst.dev))
 		ip_learn_oob_route(net, flp4, rt);
 
 	return rt;
-- 
2.30.2

